File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.cpp
#define ll long long
class Solution {
public:
    long long mostPoints(vector<vector<int>>& questions) {
        vector<ll> dp(questions.size(), -1);
        return solve(0, questions, dp);
    }
    ll solve(int index, vector<vector<int>>&questions, vector<ll> &dp){
        if(index >= questions.size()) return 0;
        if(dp[index] != -1) return dp[index];
        ll takeIt = questions[index][0] + solve(index+questions[index][1]+1, questions, dp);
        ll leaveIt = solve(index+1, questions, dp);
        return dp[index] = max(takeIt, leaveIt);
    }
};

File: 2140-solving-questions-with-brainpower/2140-solving-questions-with-brainpower.py
class Solution:
    def mostPoints(self, questions: List[List[int]]) -> int:
        dp = [0]*len(questions)
        return self.solve(0, questions, dp)
    def solve(self, index, questions, dp):
        if index >= len(questions): return 0
        if dp[index] != 0: return dp[index]
        takeIt = questions[index][0] + self.solve(index+questions[index][1]+1, questions, dp);
        leaveIt = self.solve(index+1, questions, dp)
        dp[index] = max(takeIt, leaveIt)
        return dp[index]